packages/adminrouter/extra/src/README.md

```diff
@@ -692,13 +718,229 @@ Tracking the chain of fixtures and how they use each other may provide better
 token used by Admin Router itself.
 
+### Tests code reuse and deduplication
+Lots of endpoints exposed by Admin Router share common behaviour. Usually, it
+is something simple like testing for correct headers, the right upstream, etc...
+This results in tests covering them being very repetitive.
+
+In order to keep the code DRY, common code was extracted into a library with
+generic tests and helpers and a testing framework using YAML configuration files
+was developed.
+
+#### Generic tests
+Tests shared by most of the endpoints are as follows:
+* `generic_no_slash_redirect_test` - test if a request for a location without
+  trailing slash is redirected to the one that ends with `/`
+* `generic_response_headers_verify_test` - test if the response sent by AR is
+  correct, this mostly involves checking response headers
+* `generic_upstream_headers_verify_test` - test if the request sent by NGINX to
+  the upstream is correct. Again - this is mostly about the headers now.
+* `generic_correct_upstream_dest_test` - test if the upstream request done by
+  Admin Router is sent to the correct upstream
+* `generic_correct_upstream_request_test` - test if the path component of the
+  request sent to the upstream by AR, along with HTTP version, is correct
+* `generic_location_header_during_redirect_is_adjusted_test` - test if `Location`
+  headers are being rewritten/adjusted by NGINX in a situation when the upstream
+  performs a redirect
+
+All of these tests can be found in `test-harness/modules/generic_test_code/\*.py`
+together with extensive docstrings which explain all the call parameters.
+
+#### Helpers
+Helpers shared by most of the tests are as follow:
+* `overriden_file_content` - temporarily, while in the scope of the context
+  manager, substitute the contents of the given file with a given content. At
+  the end of the context, the original file contents are restored.
+* `verify_header` and `header_is_absent` - these functions check if the given
+  list of header tuples contains or does not contain given header.
+* `GuardedSubprocess` - launches given ar instance and makes sure that it is
+  terminated at the end of the context.
+* `iam_denies_all_requests` - makes sure that within of the context, IAM denies
+  all PolicyQuery requests.
+* `ping_mesos_agent` - tests if agent with a given ID
+  (`mocker.endpoints.mesos.AGENT1_ID` by default) can be reached.
+* `repo_is_ee` - function, upon which, all remaining test harness code
+  determines repository flavour
+* `assert_iam_queried_for_uid` - asserts that within the context, IAM has been
+  queried for given UID
+* `assert_endpoint_response` - asserts response code and log messages in Admin
+  Router stderr for request against specified path
+* `LogCatcher` - see (section)[#logcatcher] for details
+
+Please check corresponding helpers' source code for more information,
+all of them are pretty well documented using docstrings.
```

Comment:
In the review here we often wanted to have more precision in the descriptions. Please make sure that the precision that you add here is also reflected in the docstrings that you refer to here.